File: generative-ai/generative-ai-example-2/MiniGame.java
import java.awt.*;
import java.awt.event.*;

import javax.swing.JOptionPane;

public class MiniGame extends Frame implements KeyListener {
    private int trexY;
    private int obstacleX;
    private boolean gameOver;
    
    public MiniGame() {
        setSize(500, 300);
        setTitle("Mini Game");
        setVisible(true);
        addKeyListener(this);
        
        trexY = 200; // initial position of the t-rex
        obstacleX = getWidth(); // initial position of the obstacle
        gameOver = false;
        
        // Start the game loop
        while (!gameOver) {
            moveObstacle();
            repaint();

            try {
                Thread.sleep(10); // adjust the delay as needed
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }
    
    public void paint(Graphics g) {
        // Draw the t-rex
        g.setColor(Color.BLUE);
        g.fillRect(100, trexY, 50, 50);
        
        // Draw the obstacle
        g.setColor(Color.BLACK);
        g.fillRect(obstacleX, 200, 30, 50);

        //Draw the ground
        g.setColor(Color.GREEN);
        g.fillRect(0, 250, 500, 50);
    }
    
    public void moveObstacle() {
        obstacleX -= 3; // adjust the speed as needed
        
        // Check for collision
        if (obstacleX <= 100 + 50 && obstacleX + 30 >= 100 && trexY + 50 >= 200) {
            gameOver = true;
            // Display game over message and restart option
            int choice = JOptionPane.showConfirmDialog(this, "Game Over! Do you want to restart?", "Game Over", JOptionPane.YES_NO_OPTION);
            if (choice == JOptionPane.YES_OPTION) {
                restartGame();
            } else {
                System.exit(0);
            }
        } else {
            // win the game
            if (obstacleX <= 100 + 50 && obstacleX + 30 >= 100 && trexY + 50 < 200) {
                gameOver = true;
                JOptionPane.showMessageDialog(this, "You win!");
                System.exit(0);
            }
        }
        
        // Reset obstacle position when it goes off the screen
        if (obstacleX + 30 <= 0) {
            obstacleX = getWidth();
        }
    }
    
    public void restartGame() {
        trexY = 200;
        obstacleX = getWidth();
        gameOver = false;
    }
    
    public void keyPressed(KeyEvent e) {
        if (e.getKeyCode() == KeyEvent.VK_SPACE) {
            jump();
        }
    }
    
    // make a function that simulates the t-rex jumping up
    public void jump() {
        // make the t-rex jump only if it's on the ground
        if (trexY == 200) {
            for (int i = 0; i < 15; i++) {
                trexY -= 5;
            }
        }
    }

    public void keyReleased(KeyEvent e) {}
    
    public void keyTyped(KeyEvent e) {}
    
    public static void main(String[] args) {
        new MiniGame();
    }
}

File: generative-ai/generative-ai-example-2/MiniGame.js
class MiniGame {
    constructor() {
        this.canvas = document.createElement('canvas');
        this.canvas.width = 500;
        this.canvas.height = 300;
        document.querySelector('section#game-panel').appendChild(this.canvas);
        this.ctx = this.canvas.getContext('2d');
        this.trexY = 200;
        this.obstacleX = this.canvas.width;
        this.gameOver = false;
        this.createTitleScreen();
    }

    createTitleScreen() {
        let titleScreen = document.createElement('div');
        titleScreen.classList.add('title-screen');
        titleScreen.innerHTML = `
            <h2>Welcome to the Mini Game</h2>
            <button class="start-button">Start</button>
        `;
        titleScreen.querySelector('.start-button').addEventListener('click', () => {
            this.playerStartsGame();
         }
        );
        document.querySelector('#game-panel').appendChild(titleScreen);
    }

    start() {
        document.addEventListener('keydown', this.keyPressed.bind(this));
        this.gameLoop();
    }

    gameLoop() {
        if (!this.gameOver) {
            this.moveObstacle();
            this.draw();
            setTimeout(() => {
                this.gameLoop();
            }, 10); 
        }
    }

    draw() {
        this.ctx.clearRect(0, 0, this.canvas.width, this.canvas.height);
        // Draw the t-rex
        this.ctx.fillStyle = 'blue';
        this.ctx.fillRect(100, this.trexY, 50, 50);

        // Draw the obstacle
        this.ctx.fillStyle = 'black';
        this.ctx.fillRect(this.obstacleX, 200, 30, 50);

        // Draw the ground
        this.ctx.fillStyle = 'green';
        this.ctx.fillRect(0, 250, 500, 50);
    }

    playerStartsGame() {
        this.start();
    }
        
    createEndScreen() {
        let endScreen = document.createElement('div');
        endScreen.classList.add('end-screen');
        endScreen.innerHTML = `
            <h2>Game Over</h2>
            <button class="restart-button">Restart</button>
        `;
        endScreen.querySelector('.restart-button').addEventListener('click', () => {
            this.restartGame();
        });
        document.querySelector('#game-panel').appendChild(endScreen);
    }
    
    createWinScreen() {
        let winScreen = document.createElement('div');
        winScreen.classList.add('win-screen');
        winScreen.innerHTML = `
            <h2>You Win!</h2>
            <button class="restart-button">Restart</button>
        `;
        winScreen.querySelector('.restart-button').addEventListener('click', () => {
            this.restartGame();
        });
        document.querySelector('#game-panel').appendChild(winScreen);
    }

    deleteEndScreen() {
        let endScreen = document.querySelector('.end-screen');
        if (endScreen) {
            endScreen.remove();
        }
    }

    deleteWinScreen() {
        let winScreen = document.querySelector('.win-screen');
        if (winScreen) {
            winScreen.remove();
        }
    }

    moveObstacle() {
        this.obstacleX -= 3; 

        if (
            this.obstacleX <= 100 + 50 &&
            this.obstacleX + 30 >= 100 &&
            this.trexY + 50 >= 200
        ) {
            this.gameOver = true;
            this.createEndScreen();
        } else {
            if (
                this.obstacleX <= 100 + 50 &&
                this.obstacleX + 30 >= 100 &&
                this.trexY + 50 < 200
            ) {
                this.gameOver = true;
                this.createWinScreen();
            }
        }

        if (this.obstacleX + 30 <= 0) {
            this.obstacleX = this.canvas.width;
        }
    }

    restartGame() {
        this.trexY = 200;
        this.obstacleX = this.canvas.width;
        this.gameOver = false;
        this.deleteEndScreen();
        this.deleteWinScreen();
    }

    keyPressed(event) {
        if (event.code === 'Space') {
            this.jump();
        }
    }

    jump() {
        if (this.trexY === 200) {
            for (let i = 0; i < 15; i++) {
                this.trexY -= 5;
            }
        }
    }
}

new MiniGame();
